In [28]:
# importamos las librerías que necesitamos

# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np

# Configuración
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

# Gestión de los warnings
# -----------------------------------------------------------------------
import warnings
warnings.filterwarnings("ignore")

### Importación de los  datasets

In [2]:
df_flight = pd.read_csv('files/Customer Flight Activity.csv') 
df_flight.head(2)

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
0,100018,2017,1,3,0,3,1521,152.0,0,0
1,100102,2017,1,10,4,14,2030,203.0,0,0


In [3]:
df_loyalty = pd.read_csv('files/Customer Loyalty History.csv') #,index_col=0
df_loyalty.head(2)

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
0,480934,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN
1,549612,Canada,Alberta,Edmonton,T3G 6Y6,Male,College,NaN,Divorced,Star,3839.61,Standard,2016,3,NaN,NaN


### Revisión inicial de estructura, tipos de datos y tamaño de cada archivo

In [4]:
df_flight.head(2)

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
0,100018,2017,1,3,0,3,1521,152.0,0,0
1,100102,2017,1,10,4,14,2030,203.0,0,0


In [5]:
df_flight.info()

<class 'pandas.DataFrame'>
RangeIndex: 405624 entries, 0 to 405623
Data columns (total 10 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Loyalty Number               405624 non-null  int64  
 1   Year                         405624 non-null  int64  
 2   Month                        405624 non-null  int64  
 3   Flights Booked               405624 non-null  int64  
 4   Flights with Companions      405624 non-null  int64  
 5   Total Flights                405624 non-null  int64  
 6   Distance                     405624 non-null  int64  
 7   Points Accumulated           405624 non-null  float64
 8   Points Redeemed              405624 non-null  int64  
 9   Dollar Cost Points Redeemed  405624 non-null  int64  
dtypes: float64(1), int64(9)
memory usage: 30.9 MB


In [ ]:
#df_flight["Points Accumulated"].sample(10)

#visualmente para ver si hay algún float y  confirmamos. 230760    148.5


214329     68.0
34633     277.0
200145      0.0
87425       0.0
182763      0.0
304275      0.0
230760    148.5
129448    349.0
117944    187.0
37019     292.0
Name: Points Accumulated, dtype: float64

In [27]:
df_flight.shape

(405624, 10)

Comentarios: 
- Para el fichero "Customer Flight Activity",  el cuál contiene los valores de los vuelos, 9 de los 10 tipos de datos son enteros, lo que puede indicar que no contiene nulos, y la columna puntos acumulados es de caracter float, podría contener nulos.
- Tamaño del fichero: 10 columnas y 405624 filas.

In [8]:
df_loyalty.head(2)

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
0,480934,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN
1,549612,Canada,Alberta,Edmonton,T3G 6Y6,Male,College,NaN,Divorced,Star,3839.61,Standard,2016,3,NaN,NaN


In [9]:
df_loyalty.info()

<class 'pandas.DataFrame'>
RangeIndex: 16737 entries, 0 to 16736
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Loyalty Number      16737 non-null  int64  
 1   Country             16737 non-null  str    
 2   Province            16737 non-null  str    
 3   City                16737 non-null  str    
 4   Postal Code         16737 non-null  str    
 5   Gender              16737 non-null  str    
 6   Education           16737 non-null  str    
 7   Salary              12499 non-null  float64
 8   Marital Status      16737 non-null  str    
 9   Loyalty Card        16737 non-null  str    
 10  CLV                 16737 non-null  float64
 11  Enrollment Type     16737 non-null  str    
 12  Enrollment Year     16737 non-null  int64  
 13  Enrollment Month    16737 non-null  int64  
 14  Cancellation Year   2067 non-null   float64
 15  Cancellation Month  2067 non-null   float64
dtypes: float64(4), 

In [10]:
df_loyalty["Cancellation Year"].unique()
#compruebo si hay coherencia con los datos y el tipo de dato. Debería de ser int.

array([  nan, 2018., 2015., 2017., 2014., 2016., 2013.])

In [11]:
df_loyalty["Cancellation Month"].unique()
#compruebo si hay coherencia con los datos y el tipo de dato. Debería de ser int.

array([nan,  1., 12.,  4.,  2.,  7., 11.,  5.,  6., 10.,  8.,  9.,  3.])

In [12]:
df_loyalty.shape


(16737, 16)

Comentarios: 
- Para el fichero "Customer Loyalty History",  el cuál contiene los perfiles de los clientes, tipos de datos variados, cuatro columnas con float, lo que puede indicar que podría contener nulos.

- Tamaño del fichero: 16 columnas y 16737 filas.

## <font color="blue">Exploración inicial Customer Flight Activity </font>

In [13]:
df_flight.head()

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
0,100018,2017,1,3,0,3,1521,152.0,0,0
1,100102,2017,1,10,4,14,2030,203.0,0,0
2,100140,2017,1,6,0,6,1200,120.0,0,0
3,100214,2017,1,0,0,0,0,0.0,0,0
4,100272,2017,1,0,0,0,0,0.0,0,0


In [14]:
df_flight.isna().sum()/df_flight.shape[0]*100

Loyalty Number                 0.0
Year                           0.0
Month                          0.0
Flights Booked                 0.0
Flights with Companions        0.0
Total Flights                  0.0
Distance                       0.0
Points Accumulated             0.0
Points Redeemed                0.0
Dollar Cost Points Redeemed    0.0
dtype: float64

In [15]:
df_flight.duplicated().sum() 

np.int64(1864)

In [16]:
df_flight[df_flight.duplicated(keep=False)].sort_values("Loyalty Number")

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
185953,101902,2017,12,0,0,0,0,0.0,0,0
185952,101902,2017,12,0,0,0,0,0.0,0,0
101447,101902,2017,7,0,0,0,0,0.0,0,0
101448,101902,2017,7,0,0,0,0,0.0,0,0
33844,101902,2017,3,0,0,0,0,0.0,0,0
...,...,...,...,...,...,...,...,...,...,...
151971,992168,2017,9,0,0,0,0,0.0,0,0
101268,992168,2017,6,0,0,0,0,0.0,0,0
354784,992168,2018,9,0,0,0,0,0.0,0,0
304080,992168,2018,6,0,0,0,0,0.0,0,0


In [17]:
df_flight[df_flight.duplicated(keep=False)].value_counts().sum()

np.int64(3712)

In [18]:
df_flight.describe(include=np.number).T #para variables numéricas

,count,mean,std,min,25%,50%,75%,max
Loyalty Number,405624.0,550037.873084,258935.286969,100018.0,326961.00,550834.0,772194.00,999986.0
Year,405624.0,2017.500000,0.500001,2017.0,2017.00,2017.5,2018.00,2018.0
Month,405624.0,6.500000,3.452057,1.0,3.75,6.5,9.25,12.0
Flights Booked,405624.0,4.115052,5.225518,0.0,0.00,1.0,8.00,21.0
Flights with Companions,405624.0,1.031805,2.076869,0.0,0.00,0.0,1.00,11.0
Total Flights,405624.0,5.146858,6.521227,0.0,0.00,1.0,10.00,32.0
Distance,405624.0,1208.880059,1433.155320,0.0,0.00,488.0,2336.00,6293.0
Points Accumulated,405624.0,123.692721,146.599831,0.0,0.00,50.0,239.00,676.5
Points Redeemed,405624.0,30.696872,125.486049,0.0,0.00,0.0,0.00,876.0
Dollar Cost Points Redeemed,405624.0,2.484503,10.150038,0.0,0.00,0.0,0.00,71.0


Variables como Flights Booked, Total Flights, Distance, Points Accumulated, Points Redeemed y Dollar Cost Points Redeemed presentan desviaciones estándar elevadas, rango de datos amplos entre min y max y la media con la mediana están muy alejadas, por lo que podemos intuir que los datos están dispersos y que los clientes son muy distintos.

In [19]:
df_flight.describe(include=str).T # me da error porque no tengo ningún dato tipo str en el df

ValueError: No columns match the specified include or exclude data types

## <font color="blue">Exploración inicial Customer Loyalty History</font>

In [ ]:
df_loyalty.head()

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
0,480934,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN
1,549612,Canada,Alberta,Edmonton,T3G 6Y6,Male,College,NaN,Divorced,Star,3839.61,Standard,2016,3,NaN,NaN
2,429460,Canada,British Columbia,Vancouver,V6E 3D9,Male,College,NaN,Single,Star,3839.75,Standard,2014,7,2018.0,1.0
3,608370,Canada,Ontario,Toronto,P1W 1K4,Male,College,NaN,Single,Star,3839.75,Standard,2013,2,NaN,NaN
4,530508,Canada,Quebec,Hull,J8Y 3Z5,Male,Bachelor,103495.0,Married,Star,3842.79,Standard,2014,10,NaN,NaN


In [ ]:
#mirar drop_duplicados con el keep. 
#df = df.drop_duplicated() // df_ac = df_ac.drop_duplicates()
# df = pd.merge(df_hi, df_ac, on='Loyalty Number', how='left')

In [ ]:
round(df_loyalty.isna().sum()/df_loyalty.shape[0]*100,2).sort_values().tail(3)
#tail de 3 para sacar solo el % de nulos de las tres columnas.

Salary                25.32
Cancellation Year     87.65
Cancellation Month    87.65
dtype: float64

In [ ]:
df_loyalty.duplicated().sum() 

np.int64(0)

In [ ]:
df_loyalty.describe(include=np.number).T #para variables numéricas

,count,mean,std,min,25%,50%,75%,max
Loyalty Number,16737.0,549735.880445,258912.132453,100018.00,326603.00,550434.00,772019.00,999986.00
Salary,12499.0,79245.609409,35008.297285,-58486.00,59246.50,73455.00,88517.50,407228.00
CLV,16737.0,7988.896536,6860.982280,1898.01,3980.84,5780.18,8940.58,83325.38
Enrollment Year,16737.0,2015.253211,1.979111,2012.00,2014.00,2015.00,2017.00,2018.00
Enrollment Month,16737.0,6.669116,3.398958,1.00,4.00,7.00,10.00,12.00
Cancellation Year,2067.0,2016.503145,1.380743,2013.00,2016.00,2017.00,2018.00,2018.00
Cancellation Month,2067.0,6.962748,3.455297,1.00,4.00,7.00,10.00,12.00


- Existe valor negativo en min de Salary, lo cual no es lógico para un ingreso anual. Puede indicar errores en registro o valores atípicos. 
    La media y mediana están relativamente cercanas, por lo que la distribución podría acercarse a simétrica. Pendiente del rango, al revisar el salario negativo.

- CLV: Rango de valores muy altos, dispersión, mediana y media alejadas.
- Enrollment year todo bien
Enrollment month rango amplio, pero es logico al tratarse del mes que va entre 1 y 12, creo que no es representativo, lo mismo pasa con las de cancellation.

In [ ]:
df_loyalty.describe(exclude=np.number).T #para variables categóricas

,count,unique,top,freq
Country,16737,1,Canada,16737
Province,16737,11,Ontario,5404
City,16737,29,Toronto,3351
Postal Code,16737,55,V6E 3D9,911
Gender,16737,2,Female,8410
Education,16737,5,Bachelor,10475
Marital Status,16737,3,Married,9735
Loyalty Card,16737,3,Star,7637
Enrollment Type,16737,2,Standard,15766


Interpretación: 
- Todos los clientes son de Canadá.
- Ligeria mayoría de mujeres 
- La mayoría con educación universitaria.
- La mayoría casada, con la tarjeta 'Star' tipo "Standard"


In [ ]:
df_loyalty["Marital Status"].unique()
#reviso los valores unicos de Marital Status

<StringArray>
['Married', 'Divorced', 'Single']
Length: 3, dtype: str

In [ ]:
df_loyalty["Enrollment Type"].unique()

<StringArray>
['Standard', '2018 Promotion']
Length: 2, dtype: str

In [ ]:
df_loyalty["Loyalty Card"].unique()

<StringArray>
['Star', 'Aurora', 'Nova']
Length: 3, dtype: str

In [ ]:
df_loyalty[df_loyalty["Loyalty Card"] == "Aurora"].head()

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
18,172755,Canada,Alberta,Edmonton,T3G 6Y6,Female,College,NaN,Single,Aurora,5303.76,2018 Promotion,2018,3,NaN,NaN
36,938981,Canada,Alberta,Edmonton,T3G 6Y6,Female,College,NaN,Single,Aurora,5337.35,2018 Promotion,2018,4,NaN,NaN
65,386082,Canada,Alberta,Edmonton,T9G 1W3,Female,Bachelor,30486.0,Divorced,Aurora,5391.97,2018 Promotion,2018,3,2018.0,11.0
69,112522,Canada,Alberta,Peace River,T9O 2W2,Male,Bachelor,49846.0,Married,Aurora,5636.74,2018 Promotion,2018,2,NaN,NaN
73,330694,Canada,Alberta,Edmonton,T3G 6Y6,Female,Bachelor,105053.0,Married,Aurora,6172.91,2018 Promotion,2018,2,NaN,NaN


In [ ]:
df_loyalty[df_loyalty["Loyalty Card"] == "Nova"].head()

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
239,384146,Canada,Alberta,Peace River,T9O 2W2,Female,High School or Below,48890.0,Married,Nova,2790.23,2018 Promotion,2018,4,2018.0,12.0
309,977693,Canada,Alberta,Calgary,T3E 2V9,Female,College,NaN,Single,Nova,2887.74,2018 Promotion,2018,2,NaN,NaN
312,443967,Canada,Alberta,Peace River,T9O 2W2,Female,High School or Below,73327.0,Single,Nova,3086.87,2018 Promotion,2018,3,NaN,NaN
319,556492,Canada,Alberta,Edmonton,T9G 1W3,Male,College,NaN,Single,Nova,3099.54,2018 Promotion,2018,3,NaN,NaN
325,855602,Canada,Alberta,Banff,T4V 1D4,Male,Bachelor,60704.0,Single,Nova,3133.50,2018 Promotion,2018,3,NaN,NaN


In [ ]:
Si están activos es que sigue en el sistema, hacer columna con True/False